### PR TITLE
GRID-400-463-475-459: Events Refactor. ContextMenu Event. Resize last col. Drag Cell Selection

### PR DESCRIPTION
**Pull request by Dwaynekj**
@joneit 
Got rid of fin-canvas events and flattened event structure
Fixed Context Menu
Fixed mouse dragging bug that was introduced when events got silence on whitespace GRID-475 and GRID-459

### CODE REVIEW

File: src/renderer/index.js [hunk 1]
```diff
@@ -533,23 +533,32 @@ var Renderer = Base.extend('Renderer', {
...
             vc = inFirstColumn ? firstColumn : vcs.find(function(vc) { return x < vc.right; }),
             vr = vrs.find(function(vr) { return y < vr.bottom; }),
-            result = null;
+            result = {fake: false};
 
-        if (vr && vc) {
-            var mousePoint = this.grid.newPoint(x - vc.left, y - vr.top),
-                cellEvent = new this.grid.behavior.CellEvent(vc.columnIndex, vr.index);
+        //default to last row and col
+        pseudoRow = vr ? false : ((vr = vrs[vrs.length - 1]) && true);
```
Comment (joneit) [inline src/renderer/index.js]: break each out into several lines with clear logic and maybe a comment or 2

**Comment by Dwaynekj** [comment]
Note to self. Rename fireSynthetic to trigger and try to make a consistent

File: src/Hypergrid.js [hunk 2]
```diff
@@ -931,7 +931,7 @@ var Hypergrid = Base.extend('Hypergrid', {
 
             this.div.appendChild(divCanvas);
 
-            var canvas = new Canvas(divCanvas, this.renderer, this.options.canvas);
+            var canvas = new Canvas(divCanvas, this, this.options.canvas);
```
Comment (Dwaynekj) [inline src/Hypergrid.js]: Done

File: src/Hypergrid.js [hunk 3]
```diff
@@ -951,6 +951,32 @@ var Hypergrid = Base.extend('Hypergrid', {
 
     /**
      * @memberOf Hypergrid#
+     * @summary Add an event listener to me.
+     * @param {string} eventName - The type of event we are interested in.
+     * @param {function} callback - The event handler.
+     */
+    addEventListener: function(eventName, callback) {
```
Comment (Dwaynekj) [inline src/Hypergrid.js]: Done

File: src/Hypergrid.js [hunk 4]
```diff
@@ -951,6 +951,32 @@ var Hypergrid = Base.extend('Hypergrid', {
...
+        var self = this;
+        var decorator = function(e) {
+            if (self.allowEventHandlers){
+                callback.call(self, e);
+            }
+        };
+        this.canvas.addEventListener(eventName, decorator);
+    },
+
+    allowEvents: function(allow){
```
Comment (Dwaynekj) [inline src/Hypergrid.js]: Done

File: src/renderer/index.js (hunk 1 above)
Comment (Dwaynekj) [inline src/renderer/index.js]: Done

File: src/lib/Canvas.js [hunk 5]
```diff
@@ -22,11 +22,12 @@ var RESIZE_POLLING_INTERVAL = 200,
     resizeInterval,
     charMap = makeCharMap();
 
-function Canvas(div, component) {
+function Canvas(div, grid) {
```
Comment (Dwaynekj) [inline src/lib/Canvas.js]: Done

File: src/features/Feature.js [hunk 6]
```diff
@@ -274,7 +274,7 @@ var Feature = Base.extend('Feature', {
      */
     isTopLeft: function(grid, event) {
         this.deprecated('isTopLeft', 'isTopLeft(grid, event) has been deprecated as of v1.2.0 in favor of event.isCellFixed. (Will be removed in a future version.)');
```
Comment (Dwaynekj) [inline src/features/Feature.js]: N/A

File: src/features/Feature.js [hunk 7]
```diff
@@ -255,7 +255,7 @@ var Feature = Base.extend('Feature', {
      */
     isFixedColumn: function(grid, event) {
         this.deprecated('isFixedColumn', 'isFixedColumn(grid, event) has been deprecated as of v1.2.0 in favor of event.isColumnFixed. (Will be removed in a future version.)');
```
Comment (Dwaynekj) [inline src/features/Feature.js]: N/A

File: src/features/Feature.js [hunk 8]
```diff
@@ -236,7 +236,7 @@ var Feature = Base.extend('Feature', {
      */
     isFixedRow: function(grid, event) {
         this.deprecated('isFixedRow', 'isFixedRow(grid, event) has been deprecated as of v1.2.0 in favor of event.isRowFixed. (Will be removed in a future version.)');
```
Comment (Dwaynekj) [inline src/features/Feature.js]: N/A

File: src/features/CellSelection.js [hunk 9]
```diff
@@ -304,11 +303,11 @@ var CellSelection = Feature.extend('CellSelection', {
     /**
      * @memberOf CellSelection.prototype
      * @param {Hypergrid} grid
-     * @param {Object} event - the event details
+     * @param {Object} eventDetails - the event details
      */
-    handleDOWN: function(grid, event) {
+    handleDOWN: function(grid, eventDetails) {
```
Comment (Dwaynekj) [inline src/features/CellSelection.js]: N/A

File: src/features/CellSelection.js [hunk 10]
```diff
@@ -58,14 +58,13 @@ var CellSelection = Feature.extend('CellSelection', {
      * @param {Object} event - the event details
      */
     handleMouseDown: function(grid, event) {
-        var dx = event.gridCell.x,
-            dy = event.dataCell.y,
-            isSelectable = grid.behavior.getCellProperty(event.dataCell.x, event.gridCell.y, 'cellSelection');
+        var dx = event.detail.gridCell.x,
```
Comment (Dwaynekj) [inline src/features/CellSelection.js]: N/A

File: src/Hypergrid.js (hunk 4 above)
Comment (Dwaynekj) [inline src/Hypergrid.js]: N/A

**Pull request closed without merging**